src/oracles/UniV3Oracle.sol

```diff
@@ -0,0 +1,111 @@
+    error AddressZero();
+
+    /// @notice Thrown when no Chainlink oracle is added for one of tokens of a deposited Uniswap V3 NFT
+    error MissingOracle();
+
+    /// @notice UniswapV3 position manager
+    INonfungiblePositionManager public immutable positionManager;
+
+    /// @notice UniswapV3 factory
+    IUniswapV3Factory public immutable factory;
+
+    /// @notice Oracle for price estimations
+    IOracle public immutable oracle;
+
+    uint256 public constant Q96 = 2**96;
+    uint256 public constant Q48 = 2**48;
+
+    /// @notice Creates a new contract
+    /// @param positionManager_ UniswapV3 position manager
+    /// @param oracle_ Oracle
+    constructor(INonfungiblePositionManager positionManager_, IOracle oracle_) {
+        if (address(positionManager_) == address(0)) {
+            revert AddressZero();
+        }
+
+        positionManager = positionManager_;
+        factory = IUniswapV3Factory(positionManager.factory());
+        oracle = oracle_;
+    }
+
+    /// @inheritdoc INFTOracle
+    function price(uint256 nft)
+        external
+        view
+        returns (
+            bool success,
+            uint256 positionAmount,
+            address pool
+        )
+    {
+        INonfungiblePositionManager.PositionInfo memory info = positionManager.positions(nft);
+        UniswapV3FeesCalculation.PositionInfo memory positionInfo = UniswapV3FeesCalculation.PositionInfo({
+            tickLower: info.tickLower,
+            tickUpper: info.tickUpper,
+            liquidity: info.liquidity,
+            feeGrowthInside0LastX128: info.feeGrowthInside0LastX128,
+            feeGrowthInside1LastX128: info.feeGrowthInside1LastX128,
+            tokensOwed0: info.tokensOwed0,
+            tokensOwed1: info.tokensOwed1
+        });
+
+        pool = factory.getPool(info.token0, info.token1, info.fee);
+
+        uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(positionInfo.tickLower);
+        uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(positionInfo.tickUpper);
+
+        uint256[2] memory tokenAmounts;
+        uint256[2] memory pricesX96;
+        {
+            bool[2] memory successOracle;
+
+            (successOracle[0], pricesX96[0]) = oracle.price(info.token0);
+            (successOracle[1], pricesX96[1]) = oracle.price(info.token1);
+
+            if (!successOracle[0] || !successOracle[1]) {
+                return (false, 0, address(0));
+            }
+
+            uint160 sqrtRatioX96 = uint160(FullMath.sqrt(FullMath.mulDiv(pricesX96[0], Q96, pricesX96[1])) * Q48);
+
+            (, int24 tick, , , , , ) = IUniswapV3Pool(pool).slot0();
+
+            (tokenAmounts[0], tokenAmounts[1]) = LiquidityAmounts.getAmountsForLiquidity(
+                sqrtRatioX96,
+                sqrtRatioAX96,
+                sqrtRatioBX96,
+                positionInfo.liquidity
+            );
+
+            (uint256 actualTokensOwed0, uint256 actualTokensOwed1) = UniswapV3FeesCalculation._calculateUniswapFees(
```

Comment:
Can we use `INonfungiblePositionManager.PositionInfo` instead of `UniswapV3FeesCalculation.PositionInfo` in this function as well?
This way you can get rid of `UniswapV3FeesCalculation.PositionInfo memory positionInfo` completely.